Exercises/NoModules/Chapter 21/Soln21_03/Soln21_03.cpp

```diff
@@ -0,0 +1,51 @@
+// Exercise 21-3 
+#include <iostream>
+#include <concepts>
+#include <span>
+#include <array>
+#include <string>
+
+/* Not requested, but just for fun, here is a concept that prescrives 
+   a complate set of operations one would require to take an average */
+template <typename T>
+concept Averagable = requires (const T x, const T y, T z, const int i)
+{
+  { x + y }  -> std::same_as<T>;
+  { x - y }  -> std::same_as<T>;
+  { z += y } -> std::same_as<T&>;
+  { z -= y } -> std::same_as<T&>;
+  { x / i }  -> std::same_as<T>;
+  { x * i }  -> std::same_as<T>;
+  { z /= i } -> std::same_as<T&>;
+  { z *= i } -> std::same_as<T&>;
+};
+
+// By default, take the N / 2'th element
+// (Note: this template covers both the case if T would not be Averageble,
+//  and the case where N % 2 == 1).
+template <typename T, size_t N>
+auto& medianOfSorted(std::span<T, N> span)
+{
+  return span[N / 2];
+}
+
+template <Averagable T, size_t N> requires (N % 2 == 0)
+auto medianOfSorted(std::span<T, N> span)
+{
+  return (span[N / 2] + span[N / 2 + 1]) / 2;
```

Comment:
this is not correct. with this, the median of the following
`std::array values_even{ 1., 2., 3., 4., 5., 6. };`
is 4.5 while it should be 3.5.
So:
`return (span[N / 2 - 1] + span[N / 2]) / 2;`
?
But then you need to be careful that N/2-1 is a valid index.